### PR TITLE
New module etcd3

### PR DESCRIPTION
**Pull request by evrardjp**
This module uses etcd3 python library to do k/v manipulation on
an etcd3 cluster using version3 api.

##### SUMMARY
This adds an etcd3 module to ansible.
It uses a native version 3 calls to modify key/values on the cluster.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
etcd3

##### ANSIBLE VERSION

devel

##### ADDITIONAL INFORMATION

None

### CODE REVIEW

**Comment by evrardjp** [comment]
Ok, I am fixing all of this tomorrow!

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: Add boilerplate 

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Copyright: (c) 2018, Ansible Project

from __future__ import absolute_import, division, print_function
__metaclass__ = type
```

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: Try/exception for checking if package `etcd3` available or not.

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: s/2.4.2.0/2.5/

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: More description will be helpful.

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: Redundant import.

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: It will be easy for debugging if we print `exp` exception to user.
Use `to_native(exp)`

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: idem

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: idem

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]:
```
state=dict(type='str', default='present', choices=['present', 'absent']),
```

File: lib/ansible/modules/clustering/etcd3.py
Comment (Akasurde) [inline lib/ansible/modules/clustering/etcd3.py]: Remove unwanted / commented code.

File: test/integration/targets/etcd3/defaults/main.yml
Comment (Akasurde) [inline test/integration/targets/etcd3/defaults/main.yml]: Line 5 to 18 can be replace using 

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Copyright: (c) 2018, Ansible Project

```

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
@evrardjp Thanks for your contribution. I see there is another [PR](https://github.com/ansible/ansible/pull/25546/) open, could you please work with author and implement this functionality ?

**Comment by evrardjp** [comment]
@Akasurde etcd has two different apis, and therefore different client libs. I thought it was good to have two different modules (etcd2 and etcd3 modules), to call natively differently.

**Comment by Akasurde** [comment]
@evrardjp No problem. 👍

**Comment by evrardjp** [comment]
Will work on it as soon as possible. I just have a large backlog, and there is no urgency for community modules inclusion in 2.5 if I am not mistaken. But I remember that the clock is ticking, don't worry :D

**Comment by evrardjp** [comment]
@Akasurde Done. Any other suggestions?

**Comment by Akasurde** [comment]
Nope. Looks good to me. Need one more person to take a look.

File: test/integration/targets/etcd3/tasks/main.yml
Comment (gundalow) [inline test/integration/targets/etcd3/tasks/main.yml]: Could you please add a blank line between tasks, will make it a little more readable

**Review by gundalow — COMMENTED** [review]
Tests look good

File: test/integration/targets/etcd3/tasks/main.yml
Comment (gundalow) [inline test/integration/targets/etcd3/tasks/main.yml]: Missing `-name: `

File: test/integration/targets/etcd3/tasks/main.yml
Comment (evrardjp) [reply test/integration/targets/etcd3/tasks/main.yml]: Good catch.

File: test/integration/targets/etcd3/tasks/main.yml
Comment (evrardjp) [reply test/integration/targets/etcd3/tasks/main.yml]: Good catch. Fixing it rn.

**Comment by evrardjp** [comment]
@gundalow Done.

**Comment by mattclay** [comment]
CI failure in the new integration tests on multiple platforms.

**Comment by evrardjp** [comment]
I will have a look later at the chain of dependencies to install per distro. Probably tomorrow.

**Comment by evrardjp** [comment]
@Akasurde you didn't dismiss your review , while you said it looks good. It has slightly changed now, could you re-review, please?

**Comment by evrardjp** [comment]
recheck

**Comment by evrardjp** [comment]
Not ready yet, doesn't respond well to network connectivity exceptions.

**Comment by evrardjp** [comment]
It seems the etcd3 running in docker containers is flaky, and prevents reliable test execution. I don't know how to fix that, because it works fine on my machine, and the code is doing what it should do: fail if it can't connect.

**Comment by evrardjp** [comment]
Passed CI tests at the second attempt. Let me close and open it again to see if that first attempt was bad luck.

**Comment by evrardjp** [comment]
Seem to work again.  Probably a temporary issue.

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
Need one more review from another person.

**Comment by evrardjp** [comment]
@flaper87 Please note this will probably be used for openstack deployment -- reference architecture now adds etcd (etcd3 according to the commit for its inclusion).

**Comment by flaper87** [comment]
not sure how I feel about the module being called `etcd3` (hate to bikeshed, sorry). Otherwise, LGTM

**Comment by flaper87** [comment]
LGTM

**Comment by evrardjp** [comment]
@flaper87 there is already an etcd module that's working with etcd2 only and a completely different api (therefore completely different libs). So you know :D

**Comment by evrardjp** [comment]
@colin-nolan @erjohnso @fabianvf @kevensen @matbu @sdoran @sgargan @supertom @treyperry Anyone here to ship it?

**Comment by Akasurde** [comment]
@evrardjp Thanks for your valuable contribution.

**Comment by evrardjp** [comment]
Anytime. Feel free to contact me personally, or our team on freenode irc #openstack-ansible.

File: test/integration/targets/etcd3/vars/RedHat.yml
Comment (afeld) [inline test/integration/targets/etcd3/vars/RedHat.yml:3]: FYI that passing in multiple packages was previously undocumented functionality for the `package` module; making a note in #34290.

**Pull request merged**
